Add support for UDC for Apollo4P family

zeonchew opened the pull request:
This PR adds support for UDC for Ambiq Apollo4p family. The boards tested are: 
- apollo4p_evb
- apollo4p_blue_kxr_evb

Followings are the USB samples tested: 
- cdc_acm
- console
- hid-keyboard
- hid-mouse
- mass

zeonchew committed:
drivers: udc_ambiq: Implemented UDC driver for apollo4p

Added implementation for udc_ambiq with its required Kconfig and
CMakelist updates.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

tmon-nordic commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,915 @@
+/* Max Packet Size for USB endpoints */
+#define EP0_MPS   64U
+#define EP_FS_MPS 64U
+#define EP_HS_MPS 512U
```
Later the code sets the iso capability on the endpoints. Are the endpoints limited in hardware to these specific values and not the maximum allowed values?

zeonchew replied:
Hi @tmon-nordic, there is a limitation that the end of transaction is not able to be identified when OUT endpoint received data that is multiple of endpoint MPS, when the data count is lower than the triggered size. Hence, the safest way to expose this now is to limit it to 64 bytes for FS and 512 bytes for HS.

jfischer-no replied:
Where is this limitation, in your shim driver, HAL driver, or controller?

zeonchew replied:
I think it can be categorized as controller limitation.

jfischer-no commented on samples/subsys/usb/hid-keyboard/boards/apollo4p_blue_kxr_evb.overlay:
```diff
@@ -0,0 +1,17 @@
+#include <zephyr/dt-bindings/input/input-event-codes.h>
```
Please add these to your board devicetree. My threshold for accepting overlays in samples is very high to keep the samples as clean as possible.

zeonchew replied:
Hi @jfischer-no, sorry for not knowing the rule for overlay. I have moved the configuration to board device tree. will take note the same for future PR.

RichardSWheatley replied:
@zeonchew a developer's personal threshold is not a rule in Zephyr.

The only rule is if the device is on the board, then add it to the board and not a sample/test overlay.
If it is not on-chip or onboard an overlay should always be acceptable practice.

zeonchew replied:
Thanks @RichardSWheatley, Thanks for the explanation. 
In this case it is indeed better to put them into board as the declaration indeed belongs to the EVB boards. 
Thank you.

jfischer-no replied:
> The only rule is if the device is on the board, then add it to the board and not a sample/test overlay. If it is not on-chip or onboard an overlay should always be acceptable practice.

If there are no buttons on the board, the overlays would not be acceptable in any case!

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,915 @@
+	void (*irq_disable_func)(const struct device *dev);
+};
+
+/* Static variables */
```
Please cleanup your code and remove all the unnecessary comments like `/* Static variables */`.

zeonchew replied:
Hi @jfischer-no, it seems my comment style is too verbose, removed all the comments that is there just to describe what a struct/function/global-variable does. Please let me know if my style is still too verbose. Thanks.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,915 @@
+{
+	LOG_DBG("Enter test mode.");
+
+	// TODO: Study and add TestMode support or remove test mode from interface
```
"Use C89-style single line comments, /*  */. The C99-style single line comment, //, is not allowed."

zeonchew replied:
Hi @jfischer-no, sorry for missing out on the porting of test mode earlier. 
The actual test mode code is now uploaded. This should no longer be an issue. 😉

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,915 @@
+
+	// TODO: Study and add TestMode support or remove test mode from interface
+
+	__ASSERT(0, "Test mode not supported");
```
Why is there ASSERT? ` * @retval -ENOTSUP Test mode is not supported` !

zeonchew replied:
Hi @jfischer-no, sorry for missing out on the porting of test mode earlier. 
The actual test mode code is now uploaded. This should no longer be an issue. 😉

jfischer-no commented on drivers/usb/udc/Kconfig.ambiq:
```diff
@@ -0,0 +1,34 @@
+ # Copyright (c) 2024 Ambiq Micro Inc.
```
What controller is it? Are there other vendors using the same controller design?

zeonchew replied:
Hi @jfischer-no, the handling in this `udc_ambiq.c` is specific to Ambiq as we need to access the USB controller functionalities via Ambiq's USB HAL. I believe that the logic here is not reusable for USB controller in other vendor's chip.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -25,6 +25,8 @@ enum udc_ambiq_event_type {
 	/* Xfer request received via udc_ambiq_ep_enqueue API */
 	UDC_AMBIQ_EVT_XFER,
+	/* ZLP Pending to be transmitted */
+	UDC_AMBIQ_EVT_ZLP_PENDING,
```
This commit should be squashed with the one that introduces the driver.
That you need a separate type for ZLP is suspicious.

zeonchew replied:
Hi @jfischer-no, Thanks for pointing this out. I was juggling on whether I should create a new event to trigger ZLP or just trigger the ZLP directly in complete event handler for IN transaction. 
Since you mentioned this as well, I think it's better to do the direct invocation instead. 
I have reverted to the said solution where there is no need for a separate `ZLP_PENDING` event

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,915 @@
+}
+
+/* Handler for USB event callbacks from USB HAL */
+static void udc_evt_callback(am_hal_usb_dev_event_e eDevState)
```
no cAmelCase

zeonchew replied:
Hi @jfischer-no, I have searched through and removed cAmelCase variable name throughout the file. Thanks for catching this 😉

aaronyegx reviewed (dismissed):
The boards and dts changes look good to me.

zeonchew commented:
Hi @jfischer-no, I believe your comments above are addressed, please help to have a review again. Thank you 😉

jfischer-no commented on boards/ambiq/apollo4p_blue_kxr_evb/apollo4p_blue_kxr_evb.dts:
```diff
@@ -45,13 +46,18 @@
 			gpios = <&gpio0_31 19 (GPIO_PULL_UP | GPIO_ACTIVE_LOW)>;
 			label = "BTN1";
+			zephyr,code = <INPUT_KEY_1>;
+			status = "okay";
```
The button changes should be in a separate commit, like all other board changes. Actually the commit sequence should look something like this: add input support to the boards, pull new manifest, dts changes, driver implementation commit, board changes to enable usb device controller.

zeonchew replied:
Hi @jfischer-no, the commit is now split with the suggested manner and sequence.

jfischer-no replied:
In the second commit, there are changes to the boards to use drivers that will be introduced in the following commit, which is not what I asked for.

zeonchew replied:
Hi @jfischer-no, looks like I did not understand correctly the intent of the original comments. 
I have changed the commit sequence to the following: 

1. Add user button to the boards
2. DTS binding
3. Driver implementation 
4. Board changes to enable USB device controller

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+	void (*irq_disable_func)(const struct device *dev);
+};
+
+const static struct device *udc_ambiq_dev;
```
How does it work if you have multiple instances?

zeonchew replied:
There is only 1 USB device controller in Ambiq chips. 
I believe that there will not be multiple instances for UDC for those chips.

jfischer-no replied:
Looks like bad HAD driver design. But variables below can be moved to udc_ambiq_data.

zeonchew replied:
Yes. You are right, that was indeed a bad driver design!
I have removed the dependency on `udc_ambiq_dev` of the functions in this driver, and moved all the variable into instance data structure `udc_ambiq_data` 

Thank you so much for pointing this out @jfischer-no

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+	return 0;
+}
+
+static void udc_evt_callback(am_hal_usb_dev_event_e dev_state)
```
Is this function called in ISR context?

zeonchew replied:
Yes, they are called in ISR context. 
Is there anything needs to be taken care since that is the case?

jfischer-no replied:
No udc_evt_callback() is fine. What else is called in the ISR context?

zeonchew replied:
Hi @jfischer-no , `udc_ep0_setup_callback()` and `udc_ep_xfer_complete_callback()` is also called in the ISR context.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	__ASSERT_NO_MSG(ep_cfg);
+	am_hal_usb_ep_stall(priv->usb_handle, ep_cfg->addr);
```
Why should disabled endpoint respond with stall handshake?

zeonchew replied:
Hi @jfischer-no, this is indeed a mistake. Thanks for catching this. It is corrected as follows: 
https://github.com/zephyrproject-rtos/zephyr/blob/ce491e0f5379c01cb5fb26bbbfe437e839b75a10/drivers/usb/udc/udc_ambiq.c#L353-L362

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+		status = am_hal_usb_ep_xfer(priv->usb_handle, ep, NULL, 0);
+	} else {
+		status = am_hal_usb_ep_xfer(priv->usb_handle, ep, buf->data, buf->len);
+	}
+	if (status != AM_HAL_STATUS_SUCCESS) {
```
```c
	}

	if (status != AM_HAL_STATUS_SUCCESS) {
```

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+			LOG_ERR("No buffer for ep 0x%02x", ep_addr);
+			udc_submit_event(dev, UDC_EVT_ERROR, -ENOBUFS);
+			return;
+		}
+		net_buf_add(buf, xfer_len);
+		evt.type = UDC_AMBIQ_EVT_HAL_OUT_CMP;
+	}
+	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
```
```c
		}

		net_buf_add(buf, xfer_len);
		evt.type = UDC_AMBIQ_EVT_HAL_OUT_CMP;
	}

	k_msgq_put(&drv_msgq, &evt, K_NO_WAIT);
```

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+		break;
+	default:
+		return -EINVAL;
+	}
+	status = am_hal_usb_ep_init(priv->usb_handle, ep_cfg->addr, endpoint_type, ep_cfg->mps);
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		LOG_ERR("am_hal_usb_ep_init failed(0x%02x), %d", ep_cfg->addr, (int)status);
+		return -EIO;
+	}
+	LOG_DBG("Enable ep 0x%02x", ep_cfg->addr);
```
```c
	}

	status = am_hal_usb_ep_init(priv->usb_handle, ep_cfg->addr, endpoint_type, ep_cfg->mps);
	if (status != AM_HAL_STATUS_SUCCESS) {
		LOG_ERR("am_hal_usb_ep_init failed(0x%02x), %d", ep_cfg->addr, (int)status);
		return -EIO;
	}

	LOG_DBG("Enable ep 0x%02x", ep_cfg->addr);
```

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,925 @@
+	ret = gpio_pin_configure_dt(&usb_vddusb33_gpio, GPIO_OUTPUT);
+	if (ret) {
+		return ret;
+	}
+	ret = gpio_pin_configure_dt(&usb_vddusb0p9_gpio, GPIO_OUTPUT);
```
```c
	}

	ret = gpio_pin_configure_dt(&usb_vddusb0p9_gpio, GPIO_OUTPUT);
```

zeonchew committed:
boards: ambiq: add user button

Added user button for apollo4p_blue_kxr_evb and apollo4p_evb

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

zeonchew committed:
dt-bindings: apollo4p: add ambiq usb binding

Added ambiq-usb bindings needed by udc_ambiq.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

zeonchew committed:
boards: ambiq: apollo4p: Add USB nodes

Add USB node to apollo4p and apollo4p_blue qualifier, and apollo4p_evb
and apollo4p_blue_kxr_evb board to enableUSB support on the MCU and
its EVB.

Signed-off-by: Chew Zeh Yang <zeon.chew@ambiq.com>

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,931 @@
+
+static int usb_power_rails_set(bool on)
+{
+#define USB_DEV_NODE DT_NODELABEL(zephyr_udc0)
+#if DT_NODE_HAS_STATUS(USB_DEV_NODE, okay) && DT_NODE_HAS_PROP(USB_DEV_NODE, vddusb33_gpios) &&    \
+	DT_NODE_HAS_PROP(USB_DEV_NODE, vddusb0p9_gpios)
```
This is not correct. (Also, you cannot use `DT_NODELABEL(zephyr_udc0)` here).
Move vddusb33_gpio and vddusb0p9_gpio to `struct udc_ambiq_config`, use `GPIO_DT_SPEC_GET_OR(DT_DRV_INST(n), reset_gpios, {0})`;

zeonchew replied:
Thanks @jfischer-no for the pointers. Added the change to use `GPIO_DT_SPEC_GET_OR` instead.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,960 @@
+	switch (dev_state) {
+	case AM_HAL_USB_DEV_EVT_BUS_RESET:
+		/* enable usb bus interrupts */
+		am_hal_usb_intr_usb_enable(priv->usb_handle,
+					   USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
+						   USB_CFG2_SuspendE_Msk | USB_CFG2_ResetE_Msk);
```
```c
		am_hal_usb_intr_usb_enable(priv->usb_handle,
					   USB_CFG2_SOFE_Msk | USB_CFG2_ResumeE_Msk |
					   USB_CFG2_SuspendE_Msk | USB_CFG2_ResetE_Msk);
```

zeonchew replied:
clang-format keeps add a tab there. I have manually reverted the change by clang-format and push. Let's see whether the CI will let us through 😄

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,960 @@
+	am_hal_usb_interrupt_service(priv->usb_handle, int_status[0], int_status[1], int_status[2]);
+}
+
+static int usb_power_rails_set(const struct device *dev, bool on)
```
The return value is never checked, maybe `static void usb_power_rails_set(const struct device *dev, bool on)`?

zeonchew replied:
Hi @jfischer-no, since the error status return might be useful to indicate that the power rail control is not successful, I have made the change to propagate the error to upper layer. Please help to have a look at the change and see whether it is okay.

jfischer-no commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,960 @@
+		return ret;
+	}
+
+	/* power rails set */
+	ret = gpio_pin_set_dt(&cfg->usb_vddusb33_gpio, on == true ? 1 : 0);
+	if (ret) {
+		return ret;
+	}
+	ret = gpio_pin_set_dt(&cfg->usb_vddusb0p9_gpio, on == true ? 1 : 0);
+	if (ret) {
+		return ret;
+	}
```
Just `on`, bool is either 0 or 1 and can be promoted to any integer type.
```c
	/* power rails set */
	ret = gpio_pin_set_dt(&cfg->usb_vddusb33_gpio, on);
```

zeonchew replied:
Hi @jfischer-no, understood and made the change. Will use implicit bool to integer casting in the future.

zeonchew commented on drivers/usb/udc/udc_ambiq.c:
```diff
@@ -0,0 +1,960 @@
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	struct gpio_dt_spec usb_vddusb33_gpio;
+	struct gpio_dt_spec usb_vddusb0p9_gpio;
```
Hi @jfischer-no, I think I have missed that point earlier. removed `usb_` from the 2 variable names
